[Bug]: DI Source Generator throws when contract or service implementation are in global namespace

### Duplicate ?

- [X] I have searched issues/discussions and did not find other issues/discussions reporting this bug.

### Product version

1.3.2 (also reproducible in 1.4.0-preview1)

### Describe expected behavior

No exception is thrown

### Describe actual behavior

An exception is thrown in the source generator.

>CSC : warning CS8785: Generator 'OperationParameterInjectionGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentException' with message 'The hintName '&lt;global namespace&gt;_IService_Operation.g.cs' contains an invalid character '&lt;' at position 0.

### Which binding

BasicHttp

### security

None

### Which .NET version

.NET 7.0.306

### Which os platform

Windows

### Code snippet used to reproduce the issue
Full repro at https://github.com/mgaffigan/CoreWcfCodeGenIssue/tree/global-namespace

```c#
using Microsoft.AspNetCore.Mvc;
using System.ServiceModel;

[ServiceContract]
public interface IService
{
    [OperationContract]
    void Operation();
}

public partial class Service : IService
{
    public void Operation([FromServices] HttpContext hc)
    {
    }
}
```


### Stacktrace if any

_No response_

--- a/src/CoreWCF.BuildTools/src/Extensions.cs
+++ b/src/CoreWCF.BuildTools/src/Extensions.cs
@@ -109,4 +109,18 @@ namespace CoreWCF.BuildTools
             return result;
         }
     }
+
+    internal static class TypedConstantExtensions
+    {
+        public static string ToSafeCSharpString(this TypedConstant typedConstant)
+        {
+            if (typedConstant.Kind == TypedConstantKind.Array)
+            {
+                return $"new [] {typedConstant.ToCSharpString()}";
+            }
+
+            return typedConstant.ToCSharpString();
+
+        }
+    }
 }

--- a/src/CoreWCF.BuildTools/src/OperationParameterInjectionGenerator.Emitter.cs
+++ b/src/CoreWCF.BuildTools/src/OperationParameterInjectionGenerator.Emitter.cs
@@ -103,6 +103,9 @@ namespace CoreWCF.BuildTools
                     _ => "internal "
                 };
 
+                bool isGlobalNamespace = operationContractSpec.ServiceContractImplementation.ContainingNamespace
+                    .IsGlobalNamespace;
+
                 string returnType = operationContractSpec.MissingOperationContract.ReturnsVoid
                     ? "void"
                     : $"{operationContractSpec.MissingOperationContract.ReturnType}";
@@ -119,9 +122,14 @@ namespace CoreWCF.BuildTools
                 _builder.Clear();
                 _builder.AppendLine($@"
 using System;
-using Microsoft.Extensions.DependencyInjection;
-namespace {operationContractSpec.ServiceContractImplementation!.ContainingNamespace}
+using Microsoft.Extensions.DependencyInjection;");
+                if (!isGlobalNamespace)
+                {
+                    _builder.AppendLine($@"namespace {operationContractSpec.ServiceContractImplementation!.ContainingNamespace}
 {{");
+                    indentor.Increment();
+                }
+
                 Stack<INamedTypeSymbol> classes = new();
                 INamedTypeSymbol containingType = operationContractSpec.ServiceContractImplementation;
                 while (containingType != null)
@@ -133,12 +141,11 @@ namespace {operationContractSpec.ServiceContractImplementation!.ContainingNamesp
                 while (classes.Count > 0)
                 {
                     containingType = classes.Pop();
-                    indentor.Increment();
                     _builder.AppendLine($@"{indentor}{GetAccessibilityModifier(containingType.DeclaredAccessibility)}partial class {containingType.Name}");
                     _builder.AppendLine($@"{indentor}{{");
+                    indentor.Increment();
                 }
 
-                indentor.Increment();
                 foreach (AttributeData attributeData in operationContractSpec.UserProvidedOperationContractImplementation.GetAttributes())
                 {
                     _builder.Append($"{indentor}[{attributeData.AttributeClass}(");
@@ -146,6 +153,7 @@ namespace {operationContractSpec.ServiceContractImplementation!.ContainingNamesp
                     _builder.Append(")]");
                     _builder.AppendLine();
                 }
+
                 _builder.AppendLine($@"{indentor}public {@async}{returnType} {operationContractSpec.MissingOperationContract.Name}({parameters})");
                 _builder.AppendLine($@"{indentor}{{");
                 indentor.Increment();
@@ -200,7 +208,8 @@ namespace {operationContractSpec.ServiceContractImplementation!.ContainingNamesp
                     _builder.AppendLine($@"{indentor}}}");
                 }
 
-                _sourceGenerationContext.AddSource(fileName, SourceText.From(_builder.ToString(), Encoding.UTF8, SourceHashAlgorithm.Sha256));
+                string sourceText = _builder.ToString();
+                _sourceGenerationContext.AddSource(fileName, SourceText.From(sourceText, Encoding.UTF8, SourceHashAlgorithm.Sha256));
 
                 string GetFileName()
                 {
@@ -213,7 +222,10 @@ namespace {operationContractSpec.ServiceContractImplementation!.ContainingNamesp
                             break;
                         }
                     }
-                    return $"{operationContractSpec.ServiceContract!.ContainingNamespace.ToDisplayString().Replace(".", "_")}_{operationContractSpec.ServiceContract.Name}_{operationContractName}.g.cs";
+
+                    return operationContractSpec.ServiceContractImplementation.ContainingNamespace.IsGlobalNamespace
+                        ? $"{operationContractSpec.ServiceContractImplementation.Name}_{operationContractName}.g.cs"
+                        : $"{operationContractSpec.ServiceContractImplementation!.ContainingNamespace.ToDisplayString().Replace(".", "_")}_{operationContractSpec.ServiceContractImplementation.Name}_{operationContractName}.g.cs";
                 }
 
                 void AppendResolveDependencies()

--- a/src/CoreWCF.BuildTools/src/OperationParameterInjectionGenerator.Emitter.cs
+++ b/src/CoreWCF.BuildTools/src/OperationParameterInjectionGenerator.Emitter.cs
@@ -73,7 +73,7 @@ namespace CoreWCF.BuildTools
 
             private void EmitOperationContract(OperationContractSpec operationContractSpec)
             {
-                string fileName = GetFileName();
+
                 var dependencies = operationContractSpec.UserProvidedOperationContractImplementation!.Parameters.Where(x => !operationContractSpec.MissingOperationContract!.Parameters.Any(p =>
                        p.IsMatchingParameter(x))).ToArray();
 
@@ -103,8 +103,10 @@ namespace CoreWCF.BuildTools
                     _ => "internal "
                 };
 
-                bool isGlobalNamespace = operationContractSpec.ServiceContractImplementation.ContainingNamespace
+                bool isServiceContractImplInGlobalNamespace = operationContractSpec.ServiceContractImplementation.ContainingNamespace
                     .IsGlobalNamespace;
+
+                string fileName = GetFileName();
 
                 string returnType = operationContractSpec.MissingOperationContract.ReturnsVoid
                     ? "void"
@@ -123,7 +125,7 @@ namespace CoreWCF.BuildTools
                 _builder.AppendLine($@"
 using System;
 using Microsoft.Extensions.DependencyInjection;");
-                if (!isGlobalNamespace)
+                if (!isServiceContractImplInGlobalNamespace)
                 {
                     _builder.AppendLine($@"namespace {operationContractSpec.ServiceContractImplementation!.ContainingNamespace}
 {{");
@@ -213,6 +215,13 @@ using Microsoft.Extensions.DependencyInjection;");
 
                 string GetFileName()
                 {
+                    StringBuilder fileNameBuilder = new();
+                    fileNameBuilder.Append(operationContractSpec.ServiceContractImplementation!.ToDisplayString().Replace(".", "_"));
+                    fileNameBuilder.Append("__");
+
+                    fileNameBuilder.Append(operationContractSpec.ServiceContract.ToDisplayString().Replace(".", "_"));
+                    fileNameBuilder.Append("_");
+
                     string operationContractName = operationContractSpec.MissingOperationContract!.Name;
                     foreach (var namedArgument in operationContractSpec.OperationContractAttributeData.NamedArguments)
                     {
@@ -223,9 +232,10 @@ using Microsoft.Extensions.DependencyInjection;");
                         }
                     }
 
-                    return operationContractSpec.ServiceContractImplementation.ContainingNamespace.IsGlobalNamespace
-                        ? $"{operationContractSpec.ServiceContractImplementation.Name}_{operationContractName}.g.cs"
-                        : $"{operationContractSpec.ServiceContractImplementation!.ContainingNamespace.ToDisplayString().Replace(".", "_")}_{operationContractSpec.ServiceContractImplementation.Name}_{operationContractName}.g.cs";
+                    fileNameBuilder.Append(operationContractName);
+                    fileNameBuilder.Append(".g.cs");
+
+                    return fileNameBuilder.ToString();
                 }
 
                 void AppendResolveDependencies()

--- a/src/CoreWCF.BuildTools/src/OperationParameterInjectionGenerator.Emitter.cs
+++ b/src/CoreWCF.BuildTools/src/OperationParameterInjectionGenerator.Emitter.cs
@@ -149,7 +149,7 @@ using Microsoft.Extensions.DependencyInjection;");
                 foreach (AttributeData attributeData in operationContractSpec.UserProvidedOperationContractImplementation.GetAttributes())
                 {
                     _builder.Append($"{indentor}[{attributeData.AttributeClass}(");
-                    _builder.Append(string.Join(", ", attributeData.ConstructorArguments.Select(x => x.ToCSharpString()).Union(attributeData.NamedArguments.Select(x => $@"{x.Key} = {x.Value.ToCSharpString()}") )));
+                    _builder.Append(string.Join(", ", attributeData.ConstructorArguments.Select(x => x.ToSafeCSharpString()).Union(attributeData.NamedArguments.Select(x => $@"{x.Key} = {x.Value.ToSafeCSharpString()}") )));
                     _builder.Append(")]");
                     _builder.AppendLine();
                 }

--- a/src/CoreWCF.BuildTools/tests/OperationParameterInjectionGeneratorTests.cs
+++ b/src/CoreWCF.BuildTools/tests/OperationParameterInjectionGeneratorTests.cs
@@ -1,7 +1,6 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System;
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
@@ -68,7 +67,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -142,7 +141,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -205,7 +204,7 @@ public partial class IdentityService : IIdentityService
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "IdentityService__IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 public partial class IdentityService
@@ -267,7 +266,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -333,7 +332,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -399,7 +398,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -424,7 +423,7 @@ namespace MyProject
     }}
 }}
 ", Encoding.UTF8, SourceHashAlgorithm.Sha256)),
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_EchoString.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_EchoString.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -490,7 +489,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -562,7 +561,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -634,7 +633,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -700,7 +699,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -725,7 +724,7 @@ namespace MyProject
     }}
 }}
 ", Encoding.UTF8, SourceHashAlgorithm.Sha256)),
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -758,7 +757,7 @@ namespace MyProject
         }
 
         [Fact]
-        public async Task FromServicesAttributeShouldWorkAsUsualForMVCControllers()
+        public async Task FromServicesAttributeShouldWorkAsUsualForMvcControllers()
         {
             var test = new VerifyGenerator.Test
             {
@@ -784,7 +783,7 @@ namespace MyProject
         }
 
         [Fact]
-        public async Task FromServicesAttributeShouldWorkAsUsualForMVCControllersWhenCombinedWithInjected()
+        public async Task FromServicesAttributeShouldWorkAsUsualForMvcControllersWhenCombinedWithInjected()
         {
             var test = new VerifyGenerator.Test
             {
@@ -869,7 +868,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -941,7 +940,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1013,7 +1012,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1094,7 +1093,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_ContainerA_ContainerB_ContainerC_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1217,7 +1216,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_Container_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1285,7 +1284,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1354,7 +1353,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1422,7 +1421,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1490,7 +1489,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1560,7 +1559,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From($@"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From($@"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1625,7 +1624,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1694,7 +1693,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1764,7 +1763,7 @@ namespace MyProject.Implementations
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_Implementations_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_Implementations_IdentityService__MyProject_Contracts_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject.Implementations
@@ -1830,7 +1829,7 @@ public partial class IdentityService : MyProject.Contracts.IIdentityService
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "IdentityService__MyProject_Contracts_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 public partial class IdentityService
@@ -1893,7 +1892,7 @@ namespace MyProject.Implementations
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_Implementations_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_Implementations_IdentityService__IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject.Implementations
@@ -1958,7 +1957,7 @@ namespace MyProject.Dummy
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_Dummy_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_Dummy_IdentityService__MyProject_Dummy_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject.Dummy
@@ -2027,7 +2026,7 @@ namespace MyProject.Dummy
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_Dummy_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_Dummy_IdentityService__MyProject_Dummy_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject.Dummy
@@ -2092,7 +2091,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -2158,7 +2157,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -2212,7 +2211,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService__MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -2314,7 +2313,6 @@ namespace MyProject
 }}
 "
                     },
-                    GeneratedSources = { },
                     ExpectedDiagnostics =
                     {
                         new DiagnosticResult(DiagnosticDescriptors.OperationParameterInjectionGenerator_01XX.ParentClassShouldBePartialError)
@@ -2362,7 +2360,6 @@ namespace MyProject
 }}
 "
                     },
-                    GeneratedSources = { },
                     ExpectedDiagnostics =
                     {
                         new DiagnosticResult(DiagnosticDescriptors.OperationParameterInjectionGenerator_01XX.ParentClassShouldBePartialError)
@@ -2413,7 +2410,6 @@ namespace MyProject
 }}
 "
                     },
-                    GeneratedSources = { },
                     ExpectedDiagnostics =
                     {
                         new DiagnosticResult(DiagnosticDescriptors.OperationParameterInjectionGenerator_01XX.ParentClassShouldBePartialError)
@@ -2456,9 +2452,7 @@ namespace MyProject
     }}
 }}
 "
-                    },
-                    GeneratedSources = { },
-                    ExpectedDiagnostics = { }
+                    }
                 }
             };
 
@@ -2497,9 +2491,7 @@ namespace MyProject
     }}
 }}
 "
-                    },
-                    GeneratedSources = { },
-                    ExpectedDiagnostics = { }
+                    }
                 }
             };
 
@@ -2541,9 +2533,7 @@ namespace MyProject
     }}
 }}
 "
-                    },
-                    GeneratedSources = { },
-                    ExpectedDiagnostics = { },
+                    }
                 }
             };
 
@@ -2578,9 +2568,7 @@ namespace MyProject
     }}
 }}
 "
-                    },
-                    GeneratedSources = { },
-                    ExpectedDiagnostics = { },
+                    }
                 }
             };
 
@@ -2590,7 +2578,7 @@ namespace MyProject
         [Theory]
         [InlineData(CoreWCFInjectedAttribute)]
         [InlineData(MVCFromServicesAttribute)]
-        public async Task ShouldNotRaiseCompilationErrorWhenParentClassDoesNotImplementOrInheritAnInterfaceWithtServiceContractAttribute(string attribute)
+        public async Task ShouldNotRaiseCompilationErrorWhenParentClassDoesNotImplementOrInheritAnInterfaceWithServiceContractAttribute(string attribute)
         {
             var test = new VerifyGenerator.Test
             {
@@ -2608,9 +2596,7 @@ namespace MyProject
     }}
 }}
 "
-                    },
-                    GeneratedSources = { },
-                    ExpectedDiagnostics = { },
+                    }
                 }
             };
 

--- a/src/CoreWCF.BuildTools/tests/OperationParameterInjectionGeneratorTests.cs
+++ b/src/CoreWCF.BuildTools/tests/OperationParameterInjectionGeneratorTests.cs
@@ -102,6 +102,81 @@ namespace MyProject
 
         [Theory]
         [MemberData(nameof(GetTestVariations))]
+        public async Task BasicTestsWithCustomAttributeContainingArrayParameter(string attributeNamespace, string attribute)
+        {
+            var test = new VerifyGenerator.Test
+            {
+                TestState =
+                {
+                    Sources =
+                    {
+@$"
+namespace MyProject
+{{
+    [AttributeUsage(AttributeTargets.Method, Inherited = false, AllowMultiple = true)]
+    public sealed class DummyAttribute : System.Attribute
+    {{
+        public DummyAttribute(string[] args)
+        {{
+        }}
+    }}
+
+    [{attributeNamespace}.ServiceContract]
+    public interface IIdentityService
+    {{
+        [{attributeNamespace}.OperationContract]
+        string Echo(string input);
+
+        [{attributeNamespace}.OperationContract]
+        string Echo2(string input);
+    }}
+
+    public partial class IdentityService : IIdentityService
+    {{
+        public string Echo(string input) => input;
+        [Dummy(new [] {{""A""}})]
+        public string Echo2(string input, [{attribute}] object a) => input;
+    }}
+}}
+"
+                    },
+                    GeneratedSources =
+                    {
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
+using System;
+using Microsoft.Extensions.DependencyInjection;
+namespace MyProject
+{{
+    public partial class IdentityService
+    {{
+        [MyProject.DummyAttribute(new [] {{""A""}})]
+        public string Echo2(string input)
+        {{
+            var serviceProvider = CoreWCF.OperationContext.Current.InstanceContext.Extensions.Find<IServiceProvider>();
+            if (serviceProvider == null) throw new InvalidOperationException(""Missing IServiceProvider in InstanceContext extensions"");
+            if (CoreWCF.OperationContext.Current.InstanceContext.IsSingleton)
+            {{
+                using (var scope = serviceProvider.CreateScope())
+                {{
+                    var d0 = scope.ServiceProvider.GetService<object>();
+                    return Echo2(input, d0);
+                }}
+            }}
+            var e0 = serviceProvider.GetService<object>();
+            return Echo2(input, e0);
+        }}
+    }}
+}}
+", Encoding.UTF8, SourceHashAlgorithm.Sha256)),
+                    },
+                },
+            };
+
+            await test.RunAsync();
+        }
+
+        [Theory]
+        [MemberData(nameof(GetTestVariations))]
         public async Task BasicTestsWithoutNamespace(string attributeNamespace, string attribute)
         {
             var test = new VerifyGenerator.Test
@@ -265,7 +340,7 @@ namespace MyProject
 {{
     public partial class IdentityService
     {{
-        [CoreWCF.OpenApi.Attributes.OpenApiResponseAttribute(ContentTypes = {{""application/json"", ""text/xml""}}, Description = ""Success"", StatusCode = System.Net.HttpStatusCode.OK, Type = typeof(string))]
+        [CoreWCF.OpenApi.Attributes.OpenApiResponseAttribute(ContentTypes = new [] {{""application/json"", ""text/xml""}}, Description = ""Success"", StatusCode = System.Net.HttpStatusCode.OK, Type = typeof(string))]
         public string Echo2(string input)
         {{
             var serviceProvider = CoreWCF.OperationContext.Current.InstanceContext.Extensions.Find<IServiceProvider>();
@@ -566,7 +641,7 @@ namespace MyProject
 {{
     public partial class IdentityService
     {{
-        [CoreWCF.AuthorizeRoleAttribute({{""Role1"", ""Role2""}})]
+        [CoreWCF.AuthorizeRoleAttribute(new [] {{""Role1"", ""Role2""}})]
         public string Echo2(string input)
         {{
             var serviceProvider = CoreWCF.OperationContext.Current.InstanceContext.Extensions.Find<IServiceProvider>();

--- a/src/CoreWCF.BuildTools/tests/OperationParameterInjectionGeneratorTests.cs
+++ b/src/CoreWCF.BuildTools/tests/OperationParameterInjectionGeneratorTests.cs
@@ -5,7 +5,6 @@ using System;
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
-using System.Reflection;
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
@@ -69,7 +68,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -103,7 +102,66 @@ namespace MyProject
 
         [Theory]
         [MemberData(nameof(GetTestVariations))]
-        public async Task BasicTestsWithMatchinSignatureButNonMatchingParemeterNames(string attributeNamespace, string attribute)
+        public async Task BasicTestsWithoutNamespace(string attributeNamespace, string attribute)
+        {
+            var test = new VerifyGenerator.Test
+            {
+                TestState =
+                {
+                    Sources =
+                    {
+@$"
+[{attributeNamespace}.ServiceContract]
+public interface IIdentityService
+{{
+    [{attributeNamespace}.OperationContract]
+    string Echo(string input);
+
+    [{attributeNamespace}.OperationContract]
+    string Echo2(string input);
+}}
+
+public partial class IdentityService : IIdentityService
+{{
+    public string Echo(string input) => input;
+    public string Echo2(string input, [{attribute}] object a) => input;
+}}
+"
+                    },
+                    GeneratedSources =
+                    {
+                        (typeof(OperationParameterInjectionGenerator), "IdentityService_Echo2.g.cs", SourceText.From(@$"
+using System;
+using Microsoft.Extensions.DependencyInjection;
+public partial class IdentityService
+{{
+    public string Echo2(string input)
+    {{
+        var serviceProvider = CoreWCF.OperationContext.Current.InstanceContext.Extensions.Find<IServiceProvider>();
+        if (serviceProvider == null) throw new InvalidOperationException(""Missing IServiceProvider in InstanceContext extensions"");
+        if (CoreWCF.OperationContext.Current.InstanceContext.IsSingleton)
+        {{
+            using (var scope = serviceProvider.CreateScope())
+            {{
+                var d0 = scope.ServiceProvider.GetService<object>();
+                return Echo2(input, d0);
+            }}
+        }}
+        var e0 = serviceProvider.GetService<object>();
+        return Echo2(input, e0);
+    }}
+}}
+", Encoding.UTF8, SourceHashAlgorithm.Sha256)),
+                    },
+                },
+            };
+
+            await test.RunAsync();
+        }
+
+        [Theory]
+        [MemberData(nameof(GetTestVariations))]
+        public async Task BasicTestsWithMatchingSignatureButNonMatchingParameterNames(string attributeNamespace, string attribute)
         {
             var test = new VerifyGenerator.Test
             {
@@ -134,7 +192,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -200,7 +258,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -266,7 +324,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -291,7 +349,7 @@ namespace MyProject
     }}
 }}
 ", Encoding.UTF8, SourceHashAlgorithm.Sha256)),
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_EchoString.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_EchoString.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -357,7 +415,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -429,7 +487,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -501,7 +559,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -567,7 +625,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -592,7 +650,7 @@ namespace MyProject
     }}
 }}
 ", Encoding.UTF8, SourceHashAlgorithm.Sha256)),
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -736,7 +794,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -808,7 +866,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -880,7 +938,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -961,7 +1019,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1084,7 +1142,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1152,7 +1210,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1221,7 +1279,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1289,7 +1347,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1357,7 +1415,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1427,7 +1485,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From($@"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From($@"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1492,7 +1550,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1561,7 +1619,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1631,7 +1689,136 @@ namespace MyProject.Implementations
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_Contracts_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_Implementations_IdentityService_Echo2.g.cs", SourceText.From(@$"
+using System;
+using Microsoft.Extensions.DependencyInjection;
+namespace MyProject.Implementations
+{{
+    public partial class IdentityService
+    {{
+        public string Echo2(string input)
+        {{
+            var serviceProvider = CoreWCF.OperationContext.Current.InstanceContext.Extensions.Find<IServiceProvider>();
+            if (serviceProvider == null) throw new InvalidOperationException(""Missing IServiceProvider in InstanceContext extensions"");
+            if (CoreWCF.OperationContext.Current.InstanceContext.IsSingleton)
+            {{
+                using (var scope = serviceProvider.CreateScope())
+                {{
+                    var d0 = scope.ServiceProvider.GetService<object>();
+                    return Echo2(input, d0);
+                }}
+            }}
+            var e0 = serviceProvider.GetService<object>();
+            return Echo2(input, e0);
+        }}
+    }}
+}}
+", Encoding.UTF8, SourceHashAlgorithm.Sha256)),
+                    },
+                },
+            };
+
+            await test.RunAsync();
+        }
+
+        [Theory]
+        [MemberData(nameof(GetTestVariations))]
+        public async Task ContractAndServiceWithDifferentNamespacesTests2(string attributeNamespace, string attribute)
+        {
+            var test = new VerifyGenerator.Test
+            {
+                TestState =
+                {
+                    Sources =
+                    {
+@$"
+namespace MyProject.Contracts
+{{
+    [{attributeNamespace}.ServiceContract]
+    public interface IIdentityService
+    {{
+        [{attributeNamespace}.OperationContract]
+        string Echo(string input);
+
+        [{attributeNamespace}.OperationContract]
+        string Echo2(string input);
+    }}
+}}
+",
+$@"
+public partial class IdentityService : MyProject.Contracts.IIdentityService
+{{
+    public string Echo(string input) => input;
+    public string Echo2(string input, [{attribute}] object a) => input;
+}}
+"
+                    },
+                    GeneratedSources =
+                    {
+                        (typeof(OperationParameterInjectionGenerator), "IdentityService_Echo2.g.cs", SourceText.From(@$"
+using System;
+using Microsoft.Extensions.DependencyInjection;
+public partial class IdentityService
+{{
+    public string Echo2(string input)
+    {{
+        var serviceProvider = CoreWCF.OperationContext.Current.InstanceContext.Extensions.Find<IServiceProvider>();
+        if (serviceProvider == null) throw new InvalidOperationException(""Missing IServiceProvider in InstanceContext extensions"");
+        if (CoreWCF.OperationContext.Current.InstanceContext.IsSingleton)
+        {{
+            using (var scope = serviceProvider.CreateScope())
+            {{
+                var d0 = scope.ServiceProvider.GetService<object>();
+                return Echo2(input, d0);
+            }}
+        }}
+        var e0 = serviceProvider.GetService<object>();
+        return Echo2(input, e0);
+    }}
+}}
+", Encoding.UTF8, SourceHashAlgorithm.Sha256)),
+                    },
+                },
+            };
+
+            await test.RunAsync();
+        }
+
+        [Theory]
+        [MemberData(nameof(GetTestVariations))]
+        public async Task ContractAndServiceWithDifferentNamespacesTests3(string attributeNamespace, string attribute)
+        {
+            var test = new VerifyGenerator.Test
+            {
+                TestState =
+                {
+                    Sources =
+                    {
+@$"
+[{attributeNamespace}.ServiceContract]
+public interface IIdentityService
+{{
+    [{attributeNamespace}.OperationContract]
+    string Echo(string input);
+
+    [{attributeNamespace}.OperationContract]
+    string Echo2(string input);
+}}
+",
+$@"
+namespace MyProject.Implementations
+{{
+    public partial class IdentityService : IIdentityService
+    {{
+        public string Echo(string input) => input;
+        public string Echo2(string input, [{attribute}] object a) => input;
+    }}
+}}
+"
+                    },
+                    GeneratedSources =
+                    {
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_Implementations_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject.Implementations
@@ -1696,7 +1883,7 @@ namespace MyProject.Dummy
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_Dummy_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_Dummy_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject.Dummy
@@ -1750,8 +1937,6 @@ namespace MyProject.Dummy
         [{attributeNamespace}.OperationContract]
         string Echo2(string input);
     }}
-
-
 }}
 ",
 @$"
@@ -1767,7 +1952,7 @@ namespace MyProject.Dummy
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_Dummy_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_Dummy_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject.Dummy
@@ -1832,7 +2017,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1898,7 +2083,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject
@@ -1952,7 +2137,7 @@ namespace MyProject
                     },
                     GeneratedSources =
                     {
-                        (typeof(OperationParameterInjectionGenerator), "MyProject_IIdentityService_Echo2.g.cs", SourceText.From(@$"
+                        (typeof(OperationParameterInjectionGenerator), "MyProject_IdentityService_Echo2.g.cs", SourceText.From(@$"
 using System;
 using Microsoft.Extensions.DependencyInjection;
 namespace MyProject

